pkg/openstack/loadbalancer.go

```diff
@@ -2567,10 +2581,64 @@ func (lbaas *LbaasV2) getSubnet(subnet string) (*subnets.Subnet, error) {
 	return nil, fmt.Errorf("found multiple subnets with name %s", subnet)
 }
 
+// ensureSecurityRule ensures to create a security rule for a defined security
+// group, if it not present.
+func (lbaas *LbaasV2) ensureSecurityRule(
+	direction rules.RuleDirection,
+	protocol rules.RuleProtocol,
+	etherType rules.RuleEtherType,
+	remoteIPPrefix, secGroupID string,
+	portRangeMin, portRangeMax int,
+) error {
+	sgListopts := rules.ListOpts{
+		Direction:      string(direction),
+		Protocol:       string(protocol),
+		PortRangeMax:   portRangeMin,
+		PortRangeMin:   portRangeMax,
+		RemoteIPPrefix: remoteIPPrefix,
+		SecGroupID:     secGroupID,
+	}
+	sgRules, err := getSecurityGroupRules(lbaas.network, sgListopts)
+	if err != nil && !cpoerrors.IsNotFound(err) {
+		return fmt.Errorf(
+			"failed to find security group rules in %s: %v", secGroupID, err)
+	}
+	if len(sgRules) != 0 {
+		return nil
+	}
+
+	sgRuleCreateOpts := rules.CreateOpts{
+		Direction:      direction,
+		Protocol:       protocol,
+		PortRangeMax:   portRangeMin,
+		PortRangeMin:   portRangeMax,
+		RemoteIPPrefix: remoteIPPrefix,
+		SecGroupID:     secGroupID,
+		EtherType:      etherType,
+	}
+
+	mc := metrics.NewMetricContext("security_group_rule", "create")
+	_, err = rules.Create(lbaas.network, sgRuleCreateOpts).Extract()
```

Comment:
@shaardie I saw the creation and update here ,maybe I missed something, those rules/sec group will be deleted 
at EnsureSecurityGroupDeleted so we get it for free?

Reply:
Yes, in  `EnsureSecurityGroupDeleted` all Rules in the Security Group are already deleted. So no need to write a function for that. 

This function to create or update Rules was extracted from the existing code and only slightly adapted. Because we create Rules now in different parts of the code (Rules for the Node Port and Rules for the Health Checks Port), I thought it was better to extract a function instead of duplicating the code.

Sorry I had directly resolved this. No intention, only clumsy fingers.

Reply:
works for me , thanks for the info :)

Reply:
@shaardie: Sorry for only noticing this when merged. How do we handle removal of SG rules in situations like a port getting edited on the .spec.ports of a Service? At a glance it seems to me a new SG rule will get created, while the old one will stay.

Reply:
I guess you are correct. I did not write any logic to remove the rules. So rules should only be deleted, if the security group itself is deleted at the moment.

Since this PR is already merged, maybe it makes sense to open a new Issue for that?

Reply:
I opened #2003.